deployment/auto_scaler.go

```diff
@@ -0,0 +1,106 @@
+package deployment
+
+import (
+	"fmt"
+	"time"
+	"v9_deployment_manager/database"
+	"v9_deployment_manager/log"
+	"v9_deployment_manager/worker"
+)
+
+type AutoScaler struct {
+	actionManager *ActionManager
+	driver        *database.Driver
+	workers       []*worker.V9Worker
+}
+
+type ComponentStatAndInstances struct {
+	numInstances int
+	averageStats worker.ComponentStats
+}
+
+var MaxHits = 100.0
+var MinHits = 15.0
+
+func (scaler *AutoScaler) AutoScale() {
+	//Get formatted worker ids
+	// TODO: Pull this out into a helper function
+	workerIDs := make([]string, len(scaler.workers))
+	for i := range scaler.workers {
+		name := fmt.Sprintf("worker_%d", i)
+		id, err := scaler.driver.FindWorkerID(name)
+		if err != nil {
+			log.Error.Println("error getting worker id:", err)
+			continue
+		}
+		workerIDs[i] = id
+	}
+
+	// Collect status of each comp on each worker
+	compMap := getCurrentInstanceState(scaler.workers)
+
+	log.Info.Println("----------------------------")
+	for _, stats := range compMap {
+		hits := stats.averageStats.Hits
+		repo := stats.averageStats.ID.Repo
+		log.Info.Println("repo: ", repo, "hits: ", hits)
+		//Evaluate if scaling up is needed
+		if stats.averageStats.Hits > MaxHits {
+			log.Info.Println("NEED MOAR POWERRR repo: ", repo)
+			scaler.actionManager.updateInstanceCount(worker.ComponentPath{
+				User: stats.averageStats.ID.User,
+				Repo: stats.averageStats.ID.Repo,
+			}, stats.numInstances+1)
+		}
+		//Evaluate if scaling down is needed
+		if stats.numInstances > 1 && stats.averageStats.Hits < MinHits {
+			//FIXME Update num instances in db
+			log.Info.Println("I shud prolly scale DOWN repo: ", repo)
+			scaler.actionManager.updateInstanceCount(worker.ComponentPath{
+				User: stats.averageStats.ID.User,
+				Repo: stats.averageStats.ID.Repo,
+			}, stats.numInstances-1)
+		}
+	}
+}
+
+func getCurrentInstanceState(workers []*worker.V9Worker) map[worker.ComponentID]*ComponentStatAndInstances {
+	// Collect status of each comp on each worker
+	compMap := make(map[worker.ComponentID]*ComponentStatAndInstances)
+	for _, w := range workers {
+		status, err := w.Status()
+		if err != nil {
+			log.Warning.Println("error getting worker status:", err)
```

Comment:
better error message plox

Reply:
How much more info would you like in this error message? The blank status? or the worker perhaps?